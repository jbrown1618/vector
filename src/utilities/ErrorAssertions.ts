import { Matrix, MatrixData } from '../types/matrix/Matrix';
import { Vector } from '../types/vector/Vector';

/**
 * Throws an error of the 2d array `data` is not square - i.e. if
 * all of its rows do not have the same length
 */
export function assertRectangular(data: MatrixData<any>): void {
  if (data.length === 0) {
    return;
  }
  const rowSize = data[0].length;
  for (let i = 0; i < data.length; i++) {
    if (data[i].length !== rowSize) {
      throw Error('TODO - message');
    }
  }
}

/**
 * Throws an error if `matrix` is not square
 */
export function assertSquare(matrix: Matrix<any>): void {
  if (matrix.getNumberOfColumns() !== matrix.getNumberOfRows()) {
    throw new Error('TODO - message');
  }
}

/**
 * Throws an error if any member of `vectors` has a different length
 * from any other member.
 */
export function assertHomogeneous(vectors: Array<Vector<any>>): void {
  if (vectors.length === 0) {
    return;
  }
  const size = vectors[0].getDimension();
  for (let i = 0; i < vectors.length; i++) {
    if (vectors[i].getDimension() !== size) {
      throw Error('TODO - message');
    }
  }
}

/**
 * Throws an error if empty
 */
export function assertNonEmpty(values: any[]) {
  if (values.length > 0) {
    return;
  }
  throw Error('TODO - message');
}

/**
 * Throws an error if `index` is not a valid identifier for an entry in `vector`
 */
export function assertValidVectorIndex(vector: Vector<any>, index: number): void {
  assertValidIndex(index, vector.getDimension());
}

/**
 * Throws an error if `(rowIndex, colIndex)` is not a valid identifier for an entry in `matrix`
 */
export function assertValidMatrixIndex(
  matrix: Matrix<any>,
  rowIndex: number,
  colIndex: number
): void {
  assertValidIndex(rowIndex, matrix.getNumberOfRows());
  assertValidIndex(colIndex, matrix.getNumberOfColumns());
}

/**
 * Throws an error if `index` is not a valid index for an array-like object of length `size`
 */
export function assertValidIndex(index: number, size: number): void {
  if (index < 0 || index >= size) {
    throw Error('TODO - message');
  }
}
