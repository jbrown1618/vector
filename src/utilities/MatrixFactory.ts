import { Matrix } from '../Matrix';
import { VectorFactory } from './VectorFactory';
import { Vector } from '../Vector';

export class MatrixFactory {
  static identity(size: number): Matrix {
    if (size < 0) {
      throw Error();
    }

    const columns: Array<Vector> = [];
    for (let i = 0; i < size; i++) {
      columns.push(VectorFactory.elementaryVector(size, i));
    }
    return Matrix.fromColumnVectors(columns);
  }
}
