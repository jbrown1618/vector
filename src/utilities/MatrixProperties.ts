import { Matrix } from '..';

export function isSquare(matrix: Matrix<any>): boolean {
  return matrix.getNumberOfColumns() === matrix.getNumberOfRows();
}
