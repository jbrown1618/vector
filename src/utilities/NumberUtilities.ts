export function approximatelyEqual(
  first: number,
  second: number,
  epsilon: number = 0.000001
): boolean {
  if (Number.isNaN(first) || Number.isNaN(second)) {
    return false; // NaN should never equal itself
  }

  if (!Number.isFinite(first) || !Number.isFinite(second)) {
    return false; // Infinities do not equal themselves
  }

  return Math.abs(first - second) < epsilon;
}

export function mod(num: number, modulus: number): number {
  // Operator % yields negative results for negative numbers.
  return ((num % modulus) + modulus) % modulus;
}

export function random(min: number = 0, max: number = 1): number {
  return min + Math.random() * (max - min);
}

export function randomNormal(
  mean: number = 0,
  standardDeviation: number = 1,
  numberOfSamples: number = 10
): number {
  let total = 0;
  for (let i = 0; i < numberOfSamples; i++) {
    total += random();
  }

  const halfNumberOfSamples = numberOfSamples / 2;
  return mean + (standardDeviation * (total - halfNumberOfSamples)) / halfNumberOfSamples;
}

// memoize these results
const binomialResults: Map<number, Map<number, number>> = new Map();
const factorialResults: Map<number, number> = new Map();
factorialResults.set(0, 1);

export function binomial(n: number, k: number): number {
  if (!binomialResults.has(n)) {
    binomialResults.set(n, new Map());
  }
  const mapForA = binomialResults.get(n) as Map<number, number>;
  if (!mapForA.has(k)) {
    if (k > n) {
      mapForA.set(k, 0);
    } else {
      const numerator = factorial(n);
      const denominator = factorial(n - k) * factorial(k);
      const result = numerator / denominator;
      mapForA.set(k, result);
    }
  }

  return mapForA.get(k) as number;
}

export function factorial(n: number): number {
  if (!factorialResults.has(n)) {
    const result = n * factorial(n - 1);
    factorialResults.set(n, result);
  }

  return factorialResults.get(n) as number;
}
