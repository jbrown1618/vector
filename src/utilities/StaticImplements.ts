/**
 * Ensures at compile time that the static part of a class implements a certain interface.
 * Usage:
 *
 * ```
 * interface HasFoo {
 *   foo(): Bar;
 * }
 *
 * @StaticImplements<HasFoo>()
 * class HasStaticFoo {
 *   static foo(): Bar;
 * }
 * ```
 */
export function StaticImplements<T>() {
  // @ts-ignore
  return (constructor: T) => {};
}
