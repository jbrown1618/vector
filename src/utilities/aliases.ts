import { Vector } from '@lib/types/vector/Vector';
import { Matrix } from '@lib/types/matrix/Matrix';
import { NumberVector } from '@lib/types/vector/NumberVector';
import { NumberMatrix } from '@lib/types/matrix/NumberMatrix';

const vb = NumberVector.builder();
const mb = NumberMatrix.builder();

/**
 * Creates a new {@link Vector} of numbers.  See {@link VectorBuilder.fromArray}
 * @public
 */
export function vec(data: number[]): Vector<number> {
  return vb.fromArray(data);
}

/**
 * Creates a new {@link Matrix} of numbers.  See {@link MatrixBuilder.fromArray}
 * @public
 */
export function mat(data: number[][]): Matrix<number> {
  return mb.fromArray(data);
}

/**
 * Creates a new identity matrix of size `size`.  See {@link MatrixBuilder.identity}
 * @public
 */
export function eye(size: number): Matrix<number> {
  return mb.identity(size);
}

/**
 * Creates a new vector of all 1s.  See {@link VectorBuilder.ones}
 * @public
 */
export function ones(entries: number): Vector<number>;
/**
 * Creates a new matrix of all 1s.  See {@link MatrixBuilder.ones}
 * @public
 */
export function ones(rows: number, columns: number): Matrix<number>;
export function ones(rows: number, columns?: number): Matrix<number> | Vector<number> {
  return columns === undefined ? vb.ones(rows) : mb.ones(rows, columns);
}

/**
 * Creates a new vector of all 0s.  See {@link VectorBuilder.zeros}
 * @public
 */
export function zeros(entries: number): Vector<number>;
/**
 * Creates a new matrix of all 0s.  See {@link MatrixBuilder.zeros}
 * @public
 */
export function zeros(rows: number, columns: number): Matrix<number>;
export function zeros(rows: number, columns?: number): Matrix<number> | Vector<number> {
  return columns === undefined ? vb.zeros(rows) : mb.zeros(rows, columns);
}

/**
 * Creates a new matrix with the specified entries on the diagonal.  See {@link MatrixBuilder.diagonal}
 * @public
 */
export function diag(elements: number[]): Matrix<number> {
  return mb.diagonal(vec(elements));
}
