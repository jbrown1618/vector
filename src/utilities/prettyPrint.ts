import { Matrix } from '../types/matrix/Matrix';
import { Vector } from '../types/vector/Vector';

/**
 * @public
 */
export function prettyPrint(num: number): string;
export function prettyPrint<S>(vector: Vector<S>): string;
export function prettyPrint<S>(matrix: Matrix<S>): string;
export function prettyPrint<S>(input: number | Vector<S> | Matrix<S>): string {
  if (typeof input === 'number') {
    return prettyPrintNumber(input);
  }
  const isVector = !!(input as any).innerProduct;
  if (isVector) {
    return prettyPrintVector(input as Vector<S>);
  } else {
    return prettyPrintMatrix(input as Matrix<S>);
  }
}

function prettyPrintNumber(num: number): string {
  const str = num.toString();
  const fixed = num.toFixed(6);
  return str.length > fixed.length ? fixed : str;
}

function prettyPrintVector<S>(vector: Vector<S>): string {
  const ops = vector.ops();
  const width = getPrintWidth(vector);
  const prettyData = vector.toArray().map(val => ops.prettyPrint(val));
  return prettyData.map(value => '[ ' + spaces(width - value.length) + value + ' ]').join('\n');
}

function prettyPrintMatrix<S>(matrix: Matrix<S>): string {
  const ops = matrix.ops();
  const widthByCol = matrix.getColumnVectors().map(col => getPrintWidth(col));
  const prettyData = matrix.toArray().map(rowArr => rowArr.map(val => ops.prettyPrint(val)));

  return prettyData
    .map(
      rowArr =>
        '[ ' +
        rowArr.map((val, colIndex) => spaces(widthByCol[colIndex] - val.length) + val).join('  ') +
        ' ]'
    )
    .join('\n');
}

function getPrintWidth<S>(v: Vector<S>): number {
  return v
    .toArray()
    .map(val => v.ops().prettyPrint(val))
    .map(str => str.length)
    .reduce((max, curr) => (curr > max ? curr : max), 0);
}

function spaces(n: number): string {
  let sp = '';
  for (let i = 0; i < n; i++) {
    sp = sp + ' ';
  }
  return sp;
}
