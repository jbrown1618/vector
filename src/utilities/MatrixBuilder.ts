import { Matrix, MatrixConstructor, MatrixData, Vector } from '..';
import { assertHomogeneous, assertValidMatrixIndex } from './ErrorAssertions';
import { ScalarOperations } from '../types/scalar/ScalarOperations';

export type MatrixIndexFunction<ScalarType> = (i: number, j: number) => ScalarType;
export type MatrixEntryFunction<ScalarType> = (
  entry: ScalarType,
  i: number,
  j: number
) => ScalarType;

export class MatrixBuilder<
  ScalarType,
  VectorType extends Vector<ScalarType>,
  MatrixType extends Matrix<ScalarType>
> {
  private readonly _matrixConstructor: MatrixConstructor<ScalarType, VectorType, MatrixType>;

  constructor(matrixConstructor: MatrixConstructor<ScalarType, VectorType, MatrixType>) {
    this._matrixConstructor = matrixConstructor;
  }

  private ops(): ScalarOperations<ScalarType> {
    return this._matrixConstructor.ops();
  }

  /**
   * Returns a 0x0 matrix
   *
   * ```
   * matrixBuilder.empty(); // []
   * ```
   */
  empty(): MatrixType {
    return new this._matrixConstructor([]);
  }

  fromData(data: MatrixData<ScalarType>): MatrixType {
    return new this._matrixConstructor(data);
  }

  fromIndexFunction(
    numRows: number,
    numColumns: number,
    indexFunction: MatrixIndexFunction<ScalarType>
  ) {
    const data: MatrixData<ScalarType> = [];
    for (let i = 0; i < numRows; i++) {
      data[i] = [];
      for (let j = 0; j < numColumns; j++) {
        data[i][j] = indexFunction(i, j);
      }
    }
    return new this._matrixConstructor(data);
  }

  map(matrix: Matrix<ScalarType>, entryFunction: MatrixEntryFunction<ScalarType>) {
    return this.fromIndexFunction(matrix.getNumberOfRows(), matrix.getNumberOfColumns(), (i, j) =>
      entryFunction(matrix.getEntry(i, j), i, j)
    );
  }

  fromColumnVectors(columns: Vector<ScalarType>[]): MatrixType {
    assertHomogeneous(columns);
    const numberOfColumns = columns.length;
    if (numberOfColumns === 0) {
      return this.empty();
    }
    const numberOfRows = columns[0].getDimension();
    if (numberOfRows === 0) {
      return this.empty();
    }

    return this.fromIndexFunction(numberOfRows, numberOfColumns, (i, j) => columns[j].getEntry(i));
  }

  fromRowVectors(rows: Vector<ScalarType>[]): MatrixType {
    assertHomogeneous(rows);
    const numberOfRows = rows.length;
    if (numberOfRows === 0) {
      return this.empty();
    }
    const numberOfColumns = rows[0].getDimension();
    if (numberOfColumns === 0) {
      return this.empty();
    }

    return this.fromIndexFunction(numberOfRows, numberOfColumns, (i, j) => rows[i].getEntry(j));
  }

  /**
   * Returns a matrix of the specified dimensions, consisting of all zeros
   *
   * ```
   * matrixBuilder.zeros(2, 3);
   *
   * // [ 0 0 0 ]
   * // [ 0 0 0 ]
   * ```
   * @param {number} numberOfRows
   * @param {number} numberOfColumns
   * @returns {MatrixType}
   */
  zeros(numberOfRows: number, numberOfColumns: number = numberOfRows): MatrixType {
    return this.fromIndexFunction(numberOfRows, numberOfColumns, () => this.ops().zero());
  }

  /**
   * Returns a matrix of the specified dimensions, consisting of all ones
   *
   * ```
   * matrixBuilder.ones(2, 3);
   *
   * // [ 1 1 1 ]
   * // [ 1 1 1 ]
   * ```
   * @param {number} numberOfRows
   * @param {number} numberOfColumns
   * @returns {MatrixType}
   */
  ones(numberOfRows: number, numberOfColumns: number = numberOfRows): MatrixType {
    return this.fromIndexFunction(numberOfRows, numberOfColumns, () => this.ops().one());
  }

  /**
   * Returns a `size` x `size` identity matrix
   *
   * ```
   * matrixBuilder.identity(3);
   *
   * // [ 1 0 0 ]
   * // [ 0 1 0 ]
   * // [ 0 0 1 ]
   * ```
   * @param {number} size
   * @returns {MatrixType}
   */
  identity(size: number): MatrixType {
    return this.fromIndexFunction(
      size,
      size,
      (i, j) => (i === j ? this.ops().one() : this.ops().zero())
    );
  }

  /**
   * Returns a square diagonal matrix whose diagonal entries come from `diagonalEntries`
   *
   * ```
   * const diagonalEntries = NumberVector.fromValues(1, 2, 3);
   * matrixBuilder.diagonal(diagonalEntries);
   *
   * // [ 1 0 0 ]
   * // [ 0 2 0 ]
   * // [ 0 0 3 ]
   * ```
   * @param {VectorType} diagonalEntries
   * @returns {MatrixType}
   */
  diagonal(diagonalEntries: VectorType): MatrixType {
    const size = diagonalEntries.getDimension();
    return this.fromIndexFunction(
      size,
      size,
      (i, j) => (i === j ? diagonalEntries.getEntry(i) : this.ops().zero())
    );
  }

  /**
   * Returns a square tridiagonal matrix whose diagonal entries correspond to the entries of
   * `diagonalEntries`, whose entries in the left-off-diagonal correspond to the entries
   * of `leftEntries`, and whose entries in the right-off-diagonal correspond fo the
   * entries of `rightEntries`.
   * The off-diagonals must have one fewer entry than the diagonal.
   * Throws an error if the dimensions are not correct.
   *
   * ```
   * const leftEntries = NumberVector.fromEntries(1, 2);
   * const diagonalEntries = NumberVector.fromEntries(3, 4, 5);
   * const rightEntries = NumberVector.fromEntries(6, 7);
   *
   * matrixBuilder.tridiagonal(leftEntries, diagonalEntries, rightEntries);
   *
   * // [ 3 6 0 ]
   * // [ 1 4 7 ]
   * // [ 0 2 5 ]
   * ```
   *
   * @param {VectorType} leftEntries
   * @param {VectorType} diagonalEntries
   * @param {VectorType} rightEntries
   * @returns {MatrixType}
   */
  tridiagonal(
    leftEntries: Vector<ScalarType>,
    diagonalEntries: Vector<ScalarType>,
    rightEntries: Vector<ScalarType>
  ): MatrixType {
    const size = diagonalEntries.getDimension();
    const hasSizeMismatch =
      leftEntries.getDimension() !== size - 1 || rightEntries.getDimension() !== size - 1;

    if (hasSizeMismatch) {
      throw Error('');
    }

    return this.fromIndexFunction(size, size, (i, j) => {
      if (i === j) {
        return diagonalEntries.getEntry(i);
      } else if (i === j + 1) {
        return leftEntries.getEntry(j);
      } else if (i === j - 1) {
        return rightEntries.getEntry(i);
      } else {
        return this.ops().zero();
      }
    });
  }

  /**
   * Returns a new matrix consisting of `left` and `right` next to one another.
   * Throws an error of `left` and `right` do not have the same number of rows.
   *
   * ```
   * const left = matrixBuilder.ones(2, 3);
   * const right = matrixBuilder.zeros(2, 4);
   *
   * matrixBuilder.augment(left, right);
   *
   * // [ 1 1 1 0 0 0 0 ]
   * // [ 1 1 1 0 0 0 0 ]
   * ```
   * @param {MatrixType} left
   * @param {MatrixType} right
   * @returns {MatrixType}
   */
  augment(left: Matrix<ScalarType>, right: Matrix<ScalarType>): MatrixType {
    if (left.getNumberOfRows() !== right.getNumberOfRows()) {
      throw Error('Dimension mismatch!');
    }

    return this.fromColumnVectors([...left.getColumnVectors(), ...right.getColumnVectors()]);
  }

  /**
   * Returns a new matrix consisting of `top` and `bottom` on top of one another.
   * Throws an error if `top` and `bottom` do not have the same number of columns.
   *
   * ```
   * const top = matrixBuilder.ones(2, 3);
   * const bottom = matrixBuilder.zeros(1,3);
   *
   * matrixBuilder.stack(top, bottom);
   *
   * // [ 1 1 1 ]
   * // [ 1 1 1 ]
   * // [ 0 0 0 ]
   * ```
   * @param {MatrixType} top
   * @param {MatrixType} bottom
   * @returns {MatrixType}
   */
  private stack(top: Matrix<ScalarType>, bottom: Matrix<ScalarType>): MatrixType {
    if (top.getNumberOfColumns() !== bottom.getNumberOfColumns()) {
      throw Error('Dimension mismatch!');
    }

    return this.fromRowVectors([...top.getRowVectors(), ...bottom.getRowVectors()]);
  }

  /**
   * Returns a single matrix consisting of a grid of matrices combined together.
   * Throws an error if any of the dimensions are incompatible.
   *
   * ```
   * const upperLeft = matrixBuilder.ones(1, 1);
   * const upperRight = matrixBuilder.zeros(1, 2);
   * const lowerLeft = matrixBuilder.zeros(2, 1);
   * const lowerRight = matrixBuilder.ones(2, 2);
   *
   * const grid = [
   *   [ upperLeft, upperRight ],
   *   [ lowerLeft, lowerRight ]
   * ];
   *
   * matrixBuilder.flatten(grid);
   *
   * // [ 1 0 0 ]
   * // [ 0 1 1 ]
   * // [ 0 1 1 ]
   * ```
   * @param {Array<Array<MatrixType>>} grid
   * @returns {MatrixType}
   */
  flatten(grid: Array<Array<MatrixType>>): MatrixType {
    if (grid.length === 0 || grid[0].length === 0) {
      return this.empty();
    }

    return grid
      .map(gridRow => {
        return gridRow.reduce((accumulator: MatrixType, gridEntry: MatrixType) => {
          return this.augment(accumulator, gridEntry);
        });
      })
      .reduce((accumulator: MatrixType, row: MatrixType) => {
        return this.stack(accumulator, row);
      });
  }

  /**
   * Returns a new matrix consisted of repetitions of a smaller matrix.
   *
   * ```
   * const I = matrixBuilder.identity(2);
   * matrixBuilder.repeat(I, 1, 2);
   *
   * // [ 1 0 1 0 ]
   * // [ 0 1 0 1 ]
   * ```
   * @param {MatrixType} matrix
   * @param {number} rows
   * @param {number} columns
   * @returns {MatrixType}
   */
  repeat(matrix: MatrixType, rows: number, columns: number) {
    const grid: Array<Array<MatrixType>> = [];

    for (let i = 0; i < rows; i++) {
      grid[i] = [];
      for (let j = 0; j < columns; j++) {
        grid[i][j] = matrix;
      }
    }

    return this.flatten(grid);
  }

  /**
   * Returns a new matrix based on a rectangular slice of a larger matrix
   *
   * ```
   * const matrix = matrixBuilder.identity(4);
   * matrixBuilder.slice(matrix, 2, 2, 3, 4);
   *
   * // [  1  0  0  0  ]
   * // [  0  1* 0* 0* ]  =>  [ 1 0 0 ]
   * // [  0  0* 1* 0* ]      [ 0 1 0 ]
   * // [  0  0  0  1  ]
   * ```
   *
   * @param {MatrixType} matrix  The original matrix
   * @param {number} rowStartIndex  The (inclusive) first row of the slice
   * @param {number} columnStartIndex  The (inclusive) first column of the slice
   * @param {number} rowEndIndex  The (exclusive) last row of the slice
   * @param {number} columnEndIndex  The (exclusive) last column of the slice
   * @returns {MatrixType}
   */
  slice(
    matrix: Matrix<ScalarType>,
    rowStartIndex: number = 0,
    columnStartIndex: number = 0,
    rowEndIndex: number = matrix.getNumberOfRows(),
    columnEndIndex: number = matrix.getNumberOfColumns()
  ): MatrixType {
    if (rowStartIndex > rowEndIndex || columnStartIndex > columnEndIndex) {
      throw Error('start index must be less than end index');
    }

    if (rowStartIndex < 0 || columnStartIndex < 0) {
      throw Error('indices must be positive');
    }

    if (rowEndIndex > matrix.getNumberOfRows() || columnEndIndex > matrix.getNumberOfColumns()) {
      throw Error('index out of bounds');
    }

    const data: MatrixData<ScalarType> = [];
    let newRowIndex = 0;
    for (let i = rowStartIndex; i < rowEndIndex; i++) {
      data[newRowIndex] = [];
      let newColumnIndex = 0;
      for (let j = columnStartIndex; j < columnEndIndex; j++) {
        data[newRowIndex][newColumnIndex] = matrix.getEntry(i, j);
        newColumnIndex++;
      }
      newRowIndex++;
    }

    return this.fromData(data);
  }

  /**
   * Returns a new matrix with all entries in row `rowToExclude` and in
   * column `columnToExclude` removed.
   *
   * ```
   * const I = matrixBuilder.identity(4);
   * matrixBuilder.slice(I, 1, 2)
   *
   * // [  1  0  0* 0  ]
   * // [  0* 1* 0* 0* ]  =>  [ 1 0 0 ]
   * // [  0  0  1* 0  ]      [ 0 0 0 ]
   * // [  0  0  0* 1  ]      [ 0 0 1 ]
   * ```
   * @param {MatrixType} matrix
   * @param {number} rowToExclude
   * @param {number} columnToExclude
   * @returns {MatrixType}
   */
  exclude(matrix: MatrixType, rowToExclude: number, columnToExclude: number): MatrixType {
    assertValidMatrixIndex(matrix, rowToExclude, columnToExclude);

    const data: MatrixData<ScalarType> = [];
    for (let i = 0; i < matrix.getNumberOfRows(); i++) {
      if (i < rowToExclude) {
        data[i] = [];
        for (let j = 0; j < matrix.getNumberOfColumns(); j++) {
          if (j < columnToExclude) {
            data[i][j] = matrix.getEntry(i, j);
          } else if (j > columnToExclude) {
            data[i][j - 1] = matrix.getEntry(i, j);
          }
        }
      } else if (i > rowToExclude) {
        data[i - 1] = [];
        for (let j = 0; j < matrix.getNumberOfColumns(); j++) {
          if (j < columnToExclude) {
            data[i - 1][j] = matrix.getEntry(i, j);
          } else if (j > columnToExclude) {
            data[i - 1][j - 1] = matrix.getEntry(i, j);
          }
        }
      }
    }

    return this.fromData(data);
  }
}
