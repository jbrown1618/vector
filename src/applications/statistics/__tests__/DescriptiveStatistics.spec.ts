import { mat, vec } from '@lib/utilities/aliases';
import {
  covariance,
  variance,
  mean,
  center,
  standardDeviation,
  correlation,
  standardize
} from '@lib/applications/statistics/DescriptiveStatistics';

describe('DescriptiveStatistics', () => {
  const sample = mat([[90, 60, 90], [90, 90, 30], [60, 60, 60], [60, 60, 90], [30, 30, 30]]);

  describe('mean', () => {
    test('calculates the mean of a vector', () => {
      expect(mean(sample.getColumn(0))).toEqual(66);
    });

    test('calculates the mean vector of a matrix', () => {
      expect(mean(sample)).toStrictEqual(vec([66, 60, 60]));
    });

    test('handles empty data', () => {
      expect(() => mean(vec([]))).toThrow();
      expect(mean(mat([]))).toStrictEqual(vec([]));
    });
  });

  describe('center', () => {
    test('centers a vector', () => {
      expect(center(sample.getColumn(0))).toStrictEqual(vec([24, 24, -6, -6, -36]));
    });

    test('centers the columns of a matrix', () => {
      expect(center(sample)).toStrictEqual(
        mat([[24, 0, 30], [24, 30, -30], [-6, 0, 0], [-6, 0, 30], [-36, -30, -30]])
      );
    });

    test('handles empty data', () => {
      expect(center(vec([]))).toStrictEqual(vec([]));
      expect(center(mat([]))).toStrictEqual(mat([]));
    });
  });

  describe('standardize', () => {
    test('centers and scales a vector', () => {
      expect(standardize(sample.getColumn(0))).toMatchSnapshot();
    });

    test('centers and standardizes the columns of a matrix', () => {
      expect(standardize(sample)).toMatchSnapshot();
    });

    test('handles empty data', () => {
      expect(standardize(vec([]))).toStrictEqual(vec([]));
      expect(standardize(mat([]))).toStrictEqual(mat([]));
    });

    test('handles homogeneous data', () => {
      expect(standardize(vec([0, 0, 0, 0, 0]))).toStrictEqual(vec([0, 0, 0, 0, 0]));
    });
  });

  describe('variance', () => {
    test('calculates the variance of a vector', () => {
      expect(variance(sample.getColumn(0))).toEqual(504);
    });

    test('calculates the variances for the columns of a matrix', () => {
      expect(variance(sample)).toStrictEqual(vec([504, 360, 720]));
    });

    test('handles empty data', () => {
      expect(() => variance(vec([]))).toThrow();
      expect(variance(mat([]))).toStrictEqual(vec([]));
    });

    test('handles homogeneous data', () => {
      expect(variance(vec([1, 1, 1, 1, 1]))).toBeCloseTo(0, 7);
    });
  });

  describe('standardDeviation', () => {
    test('calculates the standard deviation of a vector', () => {
      expect(standardDeviation(sample.getColumn(0))).toEqual(Math.sqrt(504));
    });

    test('calculates the standard deviations for the columns of a matrix', () => {
      expect(standardDeviation(sample)).toStrictEqual(
        vec([Math.sqrt(504), Math.sqrt(360), Math.sqrt(720)])
      );
    });

    test('handles empty data', () => {
      expect(() => standardDeviation(vec([]))).toThrow();
      expect(standardDeviation(mat([]))).toStrictEqual(vec([]));
    });

    test('handles homogeneous data', () => {
      expect(standardDeviation(vec([1, 1, 1, 1, 1]))).toBeCloseTo(0, 7);
    });
  });

  describe('covariance', () => {
    test('calculates the covariance of two vectors', () => {
      expect(covariance(sample.getColumn(0), sample.getColumn(1))).toEqual(360);
    });

    test('calculates the covariance matrix for a set of data', () => {
      expect(covariance(sample)).toStrictEqual(
        mat([[504, 360, 180], [360, 360, 0], [180, 0, 720]])
      );
    });

    test('handles empty data', () => {
      expect(() => covariance(vec([]), vec([]))).toThrow();
      expect(() => covariance(mat([]))).toThrow();
    });

    test('throws an error for a dimension mismatch', () => {
      expect(() => covariance(vec([1, 1, 1]), vec([1, 1, 1, 1]))).toThrow();
    });
  });

  describe('correlation', () => {
    test('calculates the correlation coefficient between two vectors', () => {
      expect(correlation(sample.getColumn(0), sample.getColumn(1))).toMatchSnapshot();
    });

    test('calculates the correlation matrix for a set of data', () => {
      expect(correlation(sample)).toMatchSnapshot();
    });

    test('handles empty data', () => {
      expect(() => correlation(vec([]), vec([]))).toThrow();
      expect(() => correlation(mat([]))).toThrow();
    });

    test('throws an error for a dimension mismatch', () => {
      expect(() => correlation(vec([1, 1, 1]), vec([1, 1, 1, 1]))).toThrow();
    });
  });
});
