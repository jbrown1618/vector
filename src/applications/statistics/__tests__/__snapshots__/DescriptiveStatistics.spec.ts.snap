// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DescriptiveStatistics correlation calculates the correlation coefficient between two vectors 1`] = `0.8451542547285167`;

exports[`DescriptiveStatistics correlation calculates the correlation matrix for a set of data 1`] = `
FloatMatrix {
  "_data": Float64Array [
    1.0000000000000002,
    0.8451542547285167,
    0.29880715233359845,
    0.8451542547285167,
    0.9999999999999999,
    0,
    0.29880715233359845,
    0,
    1.0000000000000002,
  ],
  "_shape": [
    3,
    3,
  ],
}
`;

exports[`DescriptiveStatistics standardize centers and scales a vector 1`] = `
FloatVector {
  "_data": Float64Array [
    1.0690449676496976,
    1.0690449676496976,
    -0.2672612419124244,
    -0.2672612419124244,
    -1.6035674514745464,
  ],
}
`;

exports[`DescriptiveStatistics standardize centers and standardizes the columns of a matrix 1`] = `
FloatMatrix {
  "_data": Float64Array [
    1.0690449676496976,
    1.0690449676496976,
    -0.2672612419124244,
    -0.2672612419124244,
    -1.6035674514745464,
    0,
    1.5811388300841895,
    0,
    0,
    -1.5811388300841895,
    1.118033988749895,
    -1.118033988749895,
    0,
    1.118033988749895,
    -1.118033988749895,
  ],
  "_shape": [
    5,
    3,
  ],
}
`;
